File: sdk-libs/token-pinocchio/src/instruction/transfer_checked.rs
//! TransferChecked CPI for Light Token operations.

use pinocchio::{
    AccountView as AccountInfo,
    cpi::{invoke_with_slice, invoke_signed_with_slice, Signer},
    instruction::{InstructionAccount, InstructionView},
    error::ProgramError,
    address::Address,
};

use crate::constants::LIGHT_TOKEN_PROGRAM_ID;

/// Transfer light-token checked via CPI.
///
/// # Example
///
/// ```rust,ignore
/// use light_token_pinocchio::instruction::TransferCheckedCpi;
///
/// TransferCheckedCpi {
///     source: &ctx.accounts.source,
///     mint: &ctx.accounts.mint,
///     destination: &ctx.accounts.destination,
///     amount: 100,
///     decimals: 9,
///     authority: &ctx.accounts.authority,
///     system_program: &ctx.accounts.system_program,
///     fee_payer: &ctx.accounts.fee_payer,
/// }
/// .invoke()?;
/// ```
pub struct TransferCheckedCpi<'info> {
    pub source: &'info AccountInfo,
    pub mint: &'info AccountInfo,
    pub destination: &'info AccountInfo,
    pub amount: u64,
    pub decimals: u8,
    pub authority: &'info AccountInfo,
    pub system_program: &'info AccountInfo,
    /// Fee payer for rent top-ups.
    pub fee_payer: &'info AccountInfo,
}

impl<'info> TransferCheckedCpi<'info> {
    pub fn invoke(self) -> Result<(), ProgramError> {
        self.invoke_signed(&[])
    }

    pub fn invoke_signed(self, signers: &[Signer]) -> Result<(), ProgramError> {
        let mut data = [0u8; 10]; // discriminator(1) + amount(8) + decimals(1)
        data[0] = 12u8; // TransferChecked discriminator
        data[1..9].copy_from_slice(&self.amount.to_le_bytes());
        data[9] = self.decimals;

        let program_id = Address::from(LIGHT_TOKEN_PROGRAM_ID);

        let account_metas = [
            InstructionAccount::writable(self.source.address()),
            InstructionAccount::readonly(self.mint.address()),
            InstructionAccount::writable(self.destination.address()),
            InstructionAccount::readonly_signer(self.authority.address()),
            InstructionAccount::readonly(self.system_program.address()),
            InstructionAccount::writable_signer(self.fee_payer.address()),
        ];

        let instruction = InstructionView {
            program_id: &program_id,
            accounts: &account_metas,
            data: &data,
        };

        let account_infos = [
            self.source,
            self.mint,
            self.destination,
            self.authority,
            self.system_program,
            self.fee_payer,
        ];

        if signers.is_empty() {
            invoke_with_slice(&instruction, &account_infos)
        } else {
            invoke_signed_with_slice(&instruction, &account_infos, signers)
        }
    }
}

File: js/compressed-token/src/v3/instructions/transfer-checked.ts
import {
    PublicKey,
    Signer,
    SystemProgram,
    TransactionInstruction,
} from '@solana/web3.js';
import { CTOKEN_PROGRAM_ID } from '@lightprotocol/stateless.js';
import {
    TOKEN_2022_PROGRAM_ID,
    TOKEN_PROGRAM_ID,
    createTransferCheckedInstruction as createSplTransferCheckedInstruction,
} from '@solana/spl-token';

/**
 * c-token transfer_checked instruction discriminator
 */
const CTOKEN_TRANSFER_CHECKED_DISCRIMINATOR = 12;

/**
 * Create a c-token transfer_checked instruction.
 *
 * Validates the transfer amount against the mint's decimals on-chain.
 *
 * @param source        Source c-token account
 * @param mint          Mint account (read-only, used for decimals validation)
 * @param destination   Destination c-token account
 * @param owner         Owner of the source account (signer)
 * @param amount        Amount to transfer
 * @param decimals      Expected decimals of the mint
 * @param feePayer      Optional separate fee payer for top-ups
 * @returns Transaction instruction for c-token transfer_checked
 */
export function createCTokenTransferCheckedInstruction(
    source: PublicKey,
    mint: PublicKey,
    destination: PublicKey,
    owner: PublicKey,
    amount: number | bigint,
    decimals: number,
    feePayer?: PublicKey,
): TransactionInstruction {
    // Instruction data format:
    // byte 0: discriminator (12)
    // bytes 1-8: amount (u64 LE)
    // byte 9: decimals (u8)
    const data = Buffer.alloc(10);
    data.writeUInt8(CTOKEN_TRANSFER_CHECKED_DISCRIMINATOR, 0);
    data.writeBigUInt64LE(BigInt(amount), 1);
    data.writeUInt8(decimals, 9);

    // Authority is writable when no feePayer (owner pays top-ups),
    // readonly when feePayer is provided (feePayer pays instead).
    const authorityIsWritable = !feePayer;

    const keys = [
        { pubkey: source, isSigner: false, isWritable: true },
        { pubkey: mint, isSigner: false, isWritable: false },
        { pubkey: destination, isSigner: false, isWritable: true },
        {
            pubkey: owner,
            isSigner: true,
            isWritable: authorityIsWritable,
        },
        {
            pubkey: SystemProgram.programId,
            isSigner: false,
            isWritable: false,
        },
    ];

    if (feePayer) {
        keys.push({
            pubkey: feePayer,
            isSigner: true,
            isWritable: true,
        });
    }

    return new TransactionInstruction({
        programId: CTOKEN_PROGRAM_ID,
        keys,
        data,
    });
}

/**
 * Construct a transfer_checked instruction for SPL/T22/c-token. Defaults to
 * c-token program. Validates amount against mint decimals.
 *
 * @param source        Source token account
 * @param mint          Mint account
 * @param destination   Destination token account
 * @param owner         Owner of the source account (signer)
 * @param amount        Amount to transfer
 * @param decimals      Expected decimals of the mint
 * @param multiSigners  Multi-signers (SPL only, not supported for c-token)
 * @param programId     Token program ID (default: CTOKEN_PROGRAM_ID)
 * @returns instruction for transfer_checked
 */
export function createTransferCheckedInterfaceInstruction(
    source: PublicKey,
    mint: PublicKey,
    destination: PublicKey,
    owner: PublicKey,
    amount: number | bigint,
    decimals: number,
    multiSigners: (Signer | PublicKey)[] = [],
    programId: PublicKey = CTOKEN_PROGRAM_ID,
): TransactionInstruction {
    if (programId.equals(CTOKEN_PROGRAM_ID)) {
        if (multiSigners.length > 0) {
            throw new Error(
                'c-token transfer_checked does not support multi-signers. Use a single owner.',
            );
        }
        return createCTokenTransferCheckedInstruction(
            source,
            mint,
            destination,
            owner,
            amount,
            decimals,
        );
    }

    if (
        programId.equals(TOKEN_PROGRAM_ID) ||
        programId.equals(TOKEN_2022_PROGRAM_ID)
    ) {
        return createSplTransferCheckedInstruction(
            source,
            mint,
            destination,
            owner,
            amount,
            decimals,
            multiSigners.map(pk =>
                pk instanceof PublicKey ? pk : pk.publicKey,
            ),
            programId,
        );
    }

    throw new Error(`Unsupported program ID: ${programId.toBase58()}`);
}
